gnsi/acltrie/trie.go

```diff
@@ -0,0 +1,255 @@
+		return fmt.Errorf("action unspecified")
+	}
+	if r.Mode == pathzpb.Mode_MODE_UNSPECIFIED {
+		return fmt.Errorf("mode unspecified")
+	}
+	if r.GetGroup() == "" && r.GetUser() == "" {
+		return fmt.Errorf("principal unset")
+	}
+
+	path := r.GetPath()
+	node := t.root
+
+	for _, elem := range path.Elem {
+		if elem.Name == "*" {
+			return fmt.Errorf("wildcard path names are not permitted")
+		}
+
+		// Normalize path string by pruning wildcard keys.
+		pathStr, err := elemToString(elem.Name, elem.Key)
+		if err != nil {
+			return fmt.Errorf("invalid path element: %v", err)
+		}
+
+		// If the node already exists, keep going.
+		if _, ok := node.children[pathStr]; ok {
+			node = node.children[pathStr]
+			continue
+		}
+
+		node.children[pathStr] = &trieNode{
+			elem:     elem,
+			groups:   policies{},
+			users:    policies{},
+			parent:   node,
+			children: make(map[string]*trieNode),
+		}
+
+		node = node.children[pathStr]
+	}
+	// Validate the path by ensuring that compared to all other paths with the same length and same name,
+	// all list keys do not overlap.
+	err := t.walk(func(node *trieNode, depth int) (bool, error) {
+		if depth >= len(path.Elem) {
+			return false, nil
+		}
+		if node.elem.Name != path.Elem[depth].Name {
+			return false, nil
+		}
+		if depth == len(path.Elem)-1 && node.hasPolicy {
+			pathCmp := other
+			n := node
+			for i := depth; i >= 0; i-- {
+				elemCmp, err := comparePathElem(path.Elem[i], n.elem)
+				if err != nil {
+					return false, err
+				}
+				if pathCmp != other && elemCmp != pathCmp {
+					return false, fmt.Errorf("path is not consistently subset or superset of other rules")
+				}
+				if elemCmp != other {
+					pathCmp = elemCmp
+				}
+
+				n = n.parent
+			}
+		}
+		return true, nil
+	})
+	if err != nil {
+		return fmt.Errorf("policy path conflict: %v", err)
+	}
+
+	principal := r.GetUser()
+	policy := node.users
+	if _, isUser := r.GetPrincipal().(*pathzpb.AuthorizationRule_User); !isUser {
+		principal = r.GetGroup()
+		policy = node.groups
+	}
+
+	if err := policy.insert(principal, r.GetMode(), r.GetAction()); err != nil {
+		return fmt.Errorf("error inserting policy at %v: %v", r.Path, err)
+	}
+	node.hasPolicy = true
+
+	return nil
+}
+
+// walk explores the trie in breadth first order invoke in the walkFn on every node.
+// To continue exploring children of the node, the walk func must return true.
+func (t *Trie) walk(walkFn func(node *trieNode, depth int) (bool, error)) error {
+	type traversal struct {
+		node  *trieNode
+		depth int
+	}
+
+	tr := []*traversal{{node: t.root}}
+	for len(tr) > 0 {
+		front := tr[0]
+		tr = tr[1:]
+		for _, c := range front.node.children {
+			cont, err := walkFn(c, front.depth)
+			if err != nil {
+				return err
+			}
+			if cont {
+				tr = append(tr, &traversal{node: c, depth: front.depth + 1})
+			}
+		}
+	}
+	return nil
+}
+
+type compareResult int
+
+const (
+	other compareResult = iota
+	subset
+	superset
+)
```

Comment:
I think it would make the code more readable if we split other into `equal` and `disjoint`, and then can just switch on the result. (which would incidentally make it suitable to upstream to ygot)

Reply:
agree a more generic compare utility would be nice. i've added a todo for this, because needs some extra thought to handle paths of different lengths, wildcards in the name, etc.